qemu/lib/cdrom.py

```diff
@@ -0,0 +1,162 @@
+
+    : param vm: VM object
+    : device_id: block device identifier
+    """
+    blocks_info = vm.monitor.info('block')
+
+    if isinstance(blocks_info, str):
+        open_str = 'tray open'
+        close_str = 'tray closed'
+        for block in blocks_info.splitlines():
+            if device_id in block:
+                if open_str in block:
+                    return True
+                elif close_str in block:
+                    return False
+    else:
+        for block in blocks_info:
+            if device_id in str(block) and block.get('tray_open'):
+                return block['tray_open']
+    return False
+
+
+class QMPEventCheck(object):
+    """
+    base context manager class.
+    """
+
+    def __init__(self, *args, **kargs):
+        super(QMPEventCheck, self).__init__(*args, **kargs)
+
+    def __enter__(self):
+        """
+        setup of event check.
+        """
+        raise NotImplementedError
+
+    def __exit__(self, *exc_info):
+        """
+        actual event check goes here, after the execution of event-triggering
+        function.
+        """
+        if exc_info[0]:
+            return
+        self._event_check()
+
+    def _event_check(self):
+        raise NotImplementedError
+
+
+class QMPEventCheckCD(QMPEventCheck):
+    """
+    context manager class to handle checking of event "DEVICE_TRAY_MOVED"
+    """
+
+    def __init__(self, vm, device_id, operation):
+        self.vm = vm
+        self.device_id = device_id
+        self.operation = operation
+
+    def __enter__(self):
+        """
+        initialize status of tray and clear events of qmp_monitors.
+        """
+        status = is_device_tray_opened(self.vm, self.device_id)
+        self.status_before = status
+        self.status_after = status
+
+        for m in self.vm.qmp_monitors:
+            m.clear_events()
+
+    def is_status_after_incorrect(self):
+        raise NotImplementedError
+
+    def is_events_count_incorrect(self):
+        raise NotImplementedError
+
+    @fail_on(virt_vm.VMDeviceError)
+    def _event_check(self):
+        """
+        check triggered 'DEIVCE_TRAY_MOVED' events
+
+        # 1. eject_cdrom will always open the tray and the event
+        # count is 0 if opened before, 1 if closed before.
+        # 2. change_media will always close the tray and the event
+        # count is 2 if closed before, 1 if opened before.
+        """
+        for m in self.vm.qmp_monitors:
+            events = utils_misc.wait_for(m.get_events, timeout=20)
+            if not events:
+                events = []
+            logging.info('Event list:\n%s' % events)
+            self.count = 0
+            for event in events:
+                if event['event'] == u"DEVICE_TRAY_MOVED":
+                    self.count += 1
+                    self.status_after = bool(event['data']['tray-open'])
+
+            if self.is_status_after_incorrect():
+                raise virt_vm.VMDeviceError(('Device %s tray-open '
```

Comment:
We can introduce custom exception in this module so that to not rely on `virt_vm`.